language-support/java/bindings-rxjava/src/main/java/com/daml/ledger/rxjava/components/LedgerViewFlowable.java

```diff
@@ -509,4 +509,27 @@ public int hashCode() {
             return Objects.hash(commandIdToContractIds, contractIdsToCommandIds, pendingContractSet, activeContractSet);
         }
     }
+
+    /**
+     * A ledger view for unit testing of bots.
+     *
+     * @param <R> The type of the contracts in this application.
+     */
+    public static class LedgerTestView<R> extends LedgerView {
+        public LedgerTestView(PMap<String, PMap<Identifier, PSet<String>>> commandIdToContractIds,
+                   PMap<Identifier, PMap<String, PSet<String>>> contractIdsToCommandIds,
+                   PMap<Identifier, PMap<String, R>> pendingContractSet,
+                   PMap<Identifier, PMap<String, R>> activeContractSet) {
+            super(commandIdToContractIds, contractIdsToCommandIds, pendingContractSet, activeContractSet);
+        }
+
+        @java.lang.Override
+        public LedgerView addActiveContract(Identifier templateId, String contractId, java.lang.Object object) {
```

Comment:
I think it's permitted to return a subclass of the originally specified return type in a method override.

Reply:
You can return it but you cannot modify the return type "to preserve the Subtype instead of returning the Supertype".

Reply:
Hm, so this class compiles and can be used exactly how you envisioned:

```java
    /**
     * A ledger view for unit testing of bots.
     *
     * @param <R> The type of the contracts in this application.
     */
    public static class LedgerTestView<R> extends LedgerView<R> {
        public LedgerTestView(PMap<String, PMap<Identifier, PSet<String>>> commandIdToContractIds,
                   PMap<Identifier, PMap<String, PSet<String>>> contractIdsToCommandIds,
                   PMap<Identifier, PMap<String, R>> pendingContractSet,
                   PMap<Identifier, PMap<String, R>> activeContractSet) {
            super(commandIdToContractIds, contractIdsToCommandIds, pendingContractSet, activeContractSet);
        }

        @Override
        public LedgerTestView<R> addActiveContract(Identifier templateId, String contractId, R r) {
            LedgerView lv = super.addActiveContract(templateId, contractId, r);
            return new LedgerTestView<R>(
                    lv.commandIdToContractIds,
                    lv.contractIdsToCommandIds,
                    lv.pendingContractSet,
                    lv.activeContractSet);
        }
    }

```

Reply:
Well, in the meanwhile we came to the same conclusion as written above ^ We need to modify some visibility modifiers from private to protected to allow access of those fields.

Reply:
There is a slight diff, however. We did not override the method addActiveContract but created another - we thought that way it is cleaner. Do you prefer the overridden version @gerolf-da ?

Reply:
The overridden version is just fine, because we're in Java-land after all. Especially considering that this is a class used for testing, I'm fine with that.

Reply:
Like now? :)